packages/shopping-cart/src/use-shopping-cart.ts

```diff
@@ -1,11 +1,39 @@
-import { useContext } from 'react';
-import ShoppingCartContext from './shopping-cart-context';
-import type { ShoppingCartManager } from './types';
-
-export default function useShoppingCart(): ShoppingCartManager {
-	const manager = useContext( ShoppingCartContext );
-	if ( ! manager ) {
-		throw new Error( 'useShoppingCart must be used inside a ShoppingCartProvider' );
-	}
-	return manager;
+import { useContext, useState, useEffect, useRef } from 'react';
+import { createShoppingCartHookManager } from './shopping-cart-hook-manager';
+import ShoppingCartOptionsContext from './shopping-cart-options-context';
+import useManagerClient from './use-manager-client';
+import useRefetchOnFocus from './use-refetch-on-focus';
+import type { UseShoppingCart } from './types';
+
+export default function useShoppingCart( cartKey?: string ): UseShoppingCart {
+	const managerClient = useManagerClient( 'useShoppingCart' );
+
+	const { defaultCartKey } = useContext( ShoppingCartOptionsContext );
+	const finalCartKey = cartKey ?? defaultCartKey;
+	const manager = managerClient.forCartKey( finalCartKey );
+	const hookManager = createShoppingCartHookManager( manager );
+
+	useRefetchOnFocus( finalCartKey );
+
+	useEffect( () => {
+		manager.fetchInitialCart();
+	}, [ manager ] );
+
+	const isMounted = useRef( true );
+	useEffect( () => {
+		isMounted.current = true;
+		return () => {
+			isMounted.current = false;
+		};
+	}, [] );
+
+	// Re-render when the cart changes
+	const [ cartState, setCartState ] = useState( hookManager.getState() );
+	useEffect( () => {
+		return manager.subscribe( () => {
+			isMounted.current && setCartState( hookManager.getState() );
```

Comment:
I don't understand the purpose of the `hookManager`. It returns a different value only when the underlying `manager.getState()` changes, but that state is guaranteed to change on every listener call. So it will never memoize anything.

Also, I realized that the `isMounted` check is not needed. The `manager` subscription will be cancelled on unmount, so it's guaranteed that the callback will ever be called when the component is mounted.

Reply:
> I don't understand the purpose of the hookManager. It returns a different value only when the underlying manager.getState() changes, but that state is guaranteed to change on every listener call. So it will never memoize anything.

Subscribers are notified any time an action is dispatched just after the reducer runs, but there are cases where the reducer returns the state unchanged. In those cases, the state will not have changed, so `manager.getState()` will not change, and therefore `hookManager.getState()` will not change.

> Also, I realized that the isMounted check is not needed. The manager subscription will be cancelled on unmount, so it's guaranteed that the callback will ever be called when the component is mounted.

You're right. 👍

Reply:
I changed the code so that no further actions are taken if `state === reducer( state, action )` which means that your assertion is correct and I was able to replace `hookManager.getState` with a pure function to generate the state.

Reply:
> Also, I realized that the isMounted check is not needed. The manager subscription will be cancelled on unmount, so it's guaranteed that the callback will ever be called when the component is mounted.

Actually, if I remove it then we get these errors in the checkout tests. I had to restore them.

```
   console.error
      Warning: Can't perform a React state update on an unmounted component. This is a no-op, but it indicates a memory leak in your application. To fix, cancel all subscriptions and asynchronous tasks in a useEffect cleanup function.
          in DomainContactDetails (created by ContactDetailsContainer)
```

Reply:
Digging into this some more, it appears that while the `SubscriptionManager` is notifying its subscribers, it's possible for a component to unmount and call its `useEffect` unmount callbacks; then when the subscription manager loop reaches that component and sends the notification, it tries to update its state even though the component is unmounted and has in fact already been removed from the list of subscribers. How a component has time to unmount during the notification loop is still a mystery to me but I can see it happening with debug logs during a unit test.

Reply:
This can happen when there are two nested components, each having a subscription to the store. First, `setState()` on the outer component is called, the outer component is rerendered, and the rerender means that the inner component is unmounted (because the outer JSX doesn't contain it any more). Then we call the listener of the inner component, which is already unmounted.

Redux solves this problem by organizing the subscriptions into a tree, and guaranteeing the inner ones are always called before the outer ones.

> I thought that setState (or at least its effects on components) was async

Sometimes it's async, sometimes it's sync. I don't understand these complicated rules myself ☹️